Mathlib/CategoryTheory/Localization/CalculusOfFractions/Preadditive.lean

```diff
@@ -0,0 +1,315 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Localization.CalculusOfFractions.Fractions
+import Mathlib.CategoryTheory.Localization.HasLocalization
+import Mathlib.CategoryTheory.Preadditive.AdditiveFunctor
+import Mathlib.Logic.Equiv.TransferInstance
+
+/-!
+# The preadditive category structure on the localized category
+
+In this file, it is shown that if `W : MorphismProperty C` has a left calculus
+of fractions, and `C` is preadditive, then the localized category is preadditive,
+and the localization functor is additive.
+
+Let `L : C ⥤ D` be a localization functor for `W`. We first construct an abelian
+group structure on `L.obj X ⟶ L.obj Y` for `X` and `Y` in `C`. The addition
+is defined using representatives of two morphisms in `L` as left fractions with
+the same denominator thanks to the lemmas in
+`CategoryTheory.Localization.CalculusOfFractions.Fractions`.
+As `L` is essentially surjective, we finally transport these abelian group structures
+to `X' ⟶ Y'` for all `X'` and `Y'` in `D`.
+
+Preadditive category instances are defined on the categories `W.Localization`
+(and `W.Localization'`) under the assumption the `W` has a left calculus of fractions.
+(It would be easy to deduce from the results in this file that if `W` has a right calculus
+of fractions, then the localized category can also be equipped with
+a preadditive structure, but only one of these two constructions can be made an instance!)
+
+-/
+
+namespace CategoryTheory
+
+open MorphismProperty Preadditive
+
+variable {C D : Type*} [Category C] [Category D] [Preadditive C] (L : C ⥤ D)
+  {W : MorphismProperty C} [L.IsLocalization W] [W.HasLeftCalculusOfFractions]
+
+namespace MorphismProperty
+
+/-- The opposite of a left fraction. -/
+@[reducible]
+def LeftFraction.neg {X Y : C} (φ : W.LeftFraction X Y) :
+    W.LeftFraction X Y where
+  Y' := φ.Y'
+  f := -φ.f
+  s := φ.s
+  hs := φ.hs
+
+namespace LeftFraction₂
+
+variable {X Y : C} (φ : W.LeftFraction₂ X Y)
+
+/-- The sum of two left fractions with the same denominator. -/
+@[reducible]
+def add : W.LeftFraction X Y where
+  Y' := φ.Y'
+  f := φ.f + φ.f'
+  s := φ.s
+  hs := φ.hs
+
+lemma symm_add : φ.symm.add = φ.add := by
+  dsimp [add, symm]
+  congr 1
+  apply add_comm
+
+end LeftFraction₂
+
+end MorphismProperty
+
+variable (W)
+
+namespace Localization
+
+namespace Preadditive
+
+variable {X Y Z : C}
+variable {L}
+
+/-- The opposite of a map `L.obj X ⟶ L.obj Y` when `L : C ⥤ D` is a localization
```

Comment:
I think it makes sense to wrap this whole `neg'` and `add'` section into a `section implementation_details` with a little comment saying that the whole section should never be used directly, but is only there to set up the group structure on the hom-sets.